src/query/server/query.go

```diff
@@ -541,6 +542,63 @@ func Run(runOpts RunOptions) RunResult {
 		if err != nil {
 			logger.Fatal("unable to setup downsampler for prom remote backend", zap.Error(err))
 		}
+	case config.DualStorageType:
+		// setup m3 first
+		m3dbClusters, m3dbPoolWrapper, err = initClusters(cfg, runOpts.DBConfig,
+			clusterNamespacesWatcher, runOpts.DBClient, encodingOpts,
+			runOpts.LocalSessionReadyCh, instrumentOptions,
+			tsdbOpts.CustomAdminOptions())
+		if err != nil {
+			logger.Fatal("unable to init clusters", zap.Error(err))
+		}
+
+		m3Storage, cleanup, err := newM3DBStorage(
+			cfg, m3dbClusters, m3dbPoolWrapper, queryCtxOpts, tsdbOpts, instrumentOptions,
+		)
+		if err != nil {
+			logger.Fatal("unable to setup m3db backend", zap.Error(err))
+		}
+		defer cleanup()
+
+		etcdConfig, err := resolveEtcdForM3DB(cfg)
+		if err != nil {
+			logger.Fatal("unable to resolve etcd config for m3db backend", zap.Error(err))
+		}
+
+		// setup prom-remote next
+		opts, err := promremote.NewOptions(cfg.PrometheusRemoteBackend, scope, instrumentOptions.Logger())
+		if err != nil {
+			logger.Fatal("invalid configuration", zap.Error(err))
+		}
+		promRemoteStorage, err := promremote.NewStorage(opts)
+		if err != nil {
+			logger.Fatal("unable to setup prom remote backend", zap.Error(err))
```

Comment:
Should we tolerate the setup failure here so that it won't impact writing to m3db?

Reply:
from setup perspective, i'd love to fail fast actually, but it is a good point we can tolerate remote write failures when forward